activerecord/lib/active_record/validations/numericality.rb

```diff
@@ -0,0 +1,37 @@
+# frozen_string_literal: true
+
+module ActiveRecord
+  module Validations
+    class NumericalityValidator < ActiveModel::Validations::NumericalityValidator # :nodoc:
+      def initialize(options)
+        super
+        @klass = options[:class]
+      end
+
+      def validate_each(record, attribute, value, precision: nil)
+        precision = column_precision_for(attribute) || Float::DIG
+        super
+      end
+
+      private
+        def column_precision_for(attribute)
+          if @klass < ActiveRecord::Base
+            @klass.columns_hash[attribute.to_s]&.precision
```

Comment:
Do you know if `columns_hash` takes in consideration the custom defined attributes using the `attribute` method? I think `type_for_attribute` is the correct method to take that in consideration.

Reply:
Ah, you're right, it doesn't work with virtual attributes. I've updated it to use `type_for_attribute`.